### DISCUSSION TITLE
Must every destination from the same Identity have to send an announce?

### DISCUSSION THREAD

**Discussion by ThomasFreedman** [markqvist/Reticulum#373] [Help & Questions] [answered]
Some confusion still remains on my part about paths and announces.

Currently my app creates 2 destinations from the same ID, one using RNS.Destination.IN the other with RNS.Destination.OUT. It sends announces from the OUT destination but not the IN destination.

Previously I only used one IN destination, and recipients of those announces had no trouble establishing a link back to my app. Although that worked, it seemed odd to send announces *out* from an IN destination. I also wanted to utilize the aspect filtering to replace the message type dispatch code I was using. Making sure the sender & reciever's destinations and announce handlers had matching aspects further complicated things and added some confusion.

When replaced the dispatch code I also created an OUT destination and began sending my announces from that address, and included the IN destination address in the announce app_data for recipients to use for links as before. That worked, but not sure how consistently given some of the trouble I've had at times. 

_It leads me to ask - does every single destination address, even if all are created from the same app with the same RNS.Identity, need to send an announce for a path to exist back to it?_

The reciever gets the announce, and requests a path back to the "replyTo" destination (which is an IN destination of my app, and not the OUT destination I send anounces from), but the call to RNS.Transport.request_path(replyTo.hash) never returns true. Is that b/c that IN destination doesn't send any announce? If so I am puzzled how it ever worked.

**Comment by ThomasFreedman** [comment]
Well I think I'm answering my own question here. Yes, EVERY destination MUST send an announce or a PATH will not exist to it, regardless of whether a path to another destination using the same the Identity exists.

I added another announce from the IN destination and now a path exists.

**Reply by faragher** [reply]
I guess that makes sense. While a sender could determine any given destination hash knowing the identity and aspects, the receiver could have multiple machines using the same identity, meaning destination foo.bar may have a different path (and be on a different machine in another nation!) than foo.bar.null

Thanks for running that down.

**Comment by markqvist** [comment, marked as answer]
Any `Destination` on the network must send at least one announce for it to be reachable by others. This can happen either by:

- Programatically using the `announce()` API call. See the relevant [manual section](https://markqvist.github.io/Reticulum/manual/reference.html#RNS.Destination.announce).
-  By making sure that your network is set up to recursively resolved paths. See [this section of the manual](https://markqvist.github.io/Reticulum/manual/interfaces.html#interface-modes) for that.

You **only** create a `Destination` instance with the `RNS.Destination.OUT` direction when you want to send something to that destination, that will reside somewhere else on the network. Think of it as an abstract representation of a remote `RNS.Destination.IN` instance. It's "the other end", so to speak. You put data into the `OUT` destination *representation*, and it arrives at the `IN` destination.

Don't think of this as there being two different "types" or "classes" of destinations (`IN` and `OUT`) in existence in Reticulum! There is just *destinations*, plain and simple. The only types they can have are the `SINGLE`, `GROUP` and `PLAIN` types defined in the API.

- A destination is just that, a *destination*. Something that is the *endpoint* of information.
- A `Destination` instance created with the `IN` direction is the *actual, real* destination, which holds the private keys to itself, generated from a Reticulum `Identity`.
- A `Destination` with the `OUT` direction is simply an abstract representation of the *actual* destination on the network, that you use to direct data towards it. There may exist any number of such representations on the network at any time, or there may exist none.
- You *never* send an announce from an `OUT` destination instance. In the context of Reticulum, it is meaningless.

I know I am re-hashing a bit in this post, but it's worth going over from different perspectives to make it clear ;) I know this is sometimes a bit hard to wrap your head around, if you're used to the IP world, where practically every socket is bi-directional from the start. Since there is no source addresses in Reticulum, **this is not how things work here** ;)

- Destinations are *endpoints* that can receive data.
- Once you can *reach* an endpoint, it is very easy to set up bi-directional communication, by using a `Link`, for example.
- When you have a `Link` going, you have access to the next levels of communication in the API:
  - You can easily perform `Requests` and get responses to them.
  - You can transfer arbitrary amounts of data reliably with `Resource` instances.
  - You can transport sequential data reliably with `Channel` instances.
  - You can work with reliable streams of data with `Buffer` instances.

I hope all of this makes it a lot more clear. Please read it a couple of times, it's probably necessary to grok it ;)

**Reply by ThomasFreedman** [reply]
Thank you Mark for that detailed clarification, and your patience with my slow uptake on RNS principles of operation. I keep refining my code to conform to your design philosophy. It's nearing completion, and I hope it will be a good example of how to use RNS when it is released into the wild.

I can't over emphasize how grateful I am to all you do! Thank you sir!!!

**Reply by markqvist** [reply]
Thanks for the kind words and appreciation!

I appreciate your questions. Some of these are good examples of what could be commonly misunderstood when starting out working with Reticulum, so it's highly valuable to get these points figured out and communicated efficiently :)

A lot of this stuff should go into the docs and manual when I have time to do a refresh of it all.

**Reply by ThomasFreedman** [reply]
Related questions. Thinking it out after getting it all working now:
1 - When I create an OUT destination to represent the remote endpoint, should the Identity for the remote be used or the ID of the entity creating the OUT destination?

2 - Can such a destination be used to create a link to the remote by calling the accepts_links method on it and then requesting a link? I doubt it, as that would be a security issue for the remote if external entities could "speak for" the remote by doing so.  

3 - Do destination aspects matter for creating links or sending requests? Aren't they totally unrelated to the path parameter for sending requests & registering response handlers?

4 - Isn't it as simple as matching the aspects provided to the destination used for sending announces with the aspects used to register announce handlers that receive them on the other end?

I feel very confident about the aspect parameter for distributor announce handler registration. Since announce handlers are registered thru RNS.Transport, and do not have a destination parameter it should be straight-forward aspects match between the announcer's destination aspects and the announce handler's.  This is what I found. I was confused by not realizing this and adding an erroneous aspect to the IN destination I thought the anounce handler needed.

Although the docs are clear about aspects being explicit and have no provision for wildcard matching, it didn't fully sink in. I thought by not providing any aspects to the announce handler it would see everything that matched the application name. That only works if no aspects are provided when the destination sending the announce is created.
